common/functions/src/scalars/logics/logic.rs

```diff
@@ -12,57 +12,187 @@
+    pub fn try_create(op: LogicOperator) -> Result<Box<dyn Function2>> {
+        Ok(Box::new(Self { op }))
+    }
+
+    pub fn register(factory: &mut Function2Factory) {
         factory.register("and", LogicAndFunction::desc());
         factory.register("or", LogicOrFunction::desc());
         factory.register("not", LogicNotFunction::desc());
+        factory.register("xor", LogicXorFunction::desc());
     }
 
-    pub fn try_create_func(op: DataValueLogicOperator) -> Result<Box<dyn Function>> {
-        Ok(Box::new(LogicFunction { op }))
+    fn eval_not(&self, columns: &ColumnsWithField, input_rows: usize) -> Result<ColumnRef> {
+        let mut nullable = false;
+        if columns[0].data_type().is_nullable() {
+            nullable = true;
+        }
+
+        let dt = if nullable {
+            Arc::new(NullableType::create(BooleanType::arc()))
+        } else {
+            BooleanType::arc()
+        };
+
+        let col = cast_column_field(&columns[0], &dt)?;
+
+        if nullable {
+            let col_viewer = ColumnViewer::<bool>::create(&col)?;
+
+            let mut builder = NullableColumnBuilder::<bool>::with_capacity(input_rows);
+
+            for idx in 0..input_rows {
+                builder.append(!col_viewer.value(idx), col_viewer.valid_at(idx));
+            }
+
+            Ok(builder.build(input_rows))
+        } else {
+            let col_viewer = ColumnViewer::<bool>::create(&col)?;
+            let mut builder = ColumnBuilder::<bool>::with_capacity(input_rows);
+
+            for idx in 0..input_rows {
+                builder.append(!col_viewer.value(idx));
+            }
+
+            Ok(builder.build(input_rows))
+        }
+    }
+
+    fn eval_and_not_or(&self, columns: &ColumnsWithField, input_rows: usize) -> Result<ColumnRef> {
+        let mut nullable = false;
+        if columns[0].data_type().is_nullable() || columns[1].data_type().is_nullable() {
+            nullable = true;
+        }
+
+        let dt = if nullable {
+            Arc::new(NullableType::create(BooleanType::arc()))
+        } else {
+            BooleanType::arc()
+        };
+
+        let lhs = cast_column_field(&columns[0], &dt)?;
+        let rhs = cast_column_field(&columns[1], &dt)?;
+
+        if nullable {
+            let lhs_viewer = ColumnViewer::<bool>::create(&lhs)?;
+            let rhs_viewer = ColumnViewer::<bool>::create(&rhs)?;
+
+            let mut builder = NullableColumnBuilder::<bool>::with_capacity(input_rows);
+
+            let func = match self.op {
+                LogicOperator::And => LogicAndFunction::eval_with_null,
+                LogicOperator::Or => LogicOrFunction::eval_with_null,
+                LogicOperator::Xor => LogicXorFunction::eval_with_null,
+                LogicOperator::Not => return Err(ErrorCode::LogicalError("never happen")),
+            };
+
+            for idx in 0..input_rows {
+                let (val, valid) = func(
```

Comment:
can we make this func `inline` to have better performence? cc @sundy-li

Reply:
Yes, better to test it vs before.

BTW, I do think `func` is dynamically dispatched without optimization.


I think it may be better to apply:

```
macro_rules! xxx;

match op {
LogicOperator::And => xxx!{},
...
}
```